src/core/src/sys/config/configuration.rs

```diff
@@ -0,0 +1,693 @@
+    /// ```
+    /// 0 1 6 7  2 3  4 5  # Old indexes
+    /// H-H-H-H  H-H  H-H
+    /// 0 1 2 3  4 5  6 7  # New indexes
+    /// ```
+    ///
+    /// This functions return the change in index of the first particle of the
+    /// moved molecule, i.e. in this example `4`.
+    fn merge_molecules(&mut self, first: usize, second: usize) -> usize {
+        // Do not try to merge a molecule with itself
+        if first == second {
+            return 0;
+        }
+
+        let (new_molid, old_molid) = if first < second {
+            (first, second)
+        } else {
+            (second, first)
+        };
+
+        let mut new_mol = self.molecules[new_molid].clone();
+        let old_mol = self.molecules[old_molid].clone();
+
+        if new_mol.end() == old_mol.start() {
+            // Just update the molecules ids
+            for i in old_mol {
+                self.molids[i] = new_molid;
+            }
+        } else {
+            // Move the particles close together
+            let mut new_index = new_mol.end();
+            for i in old_mol {
+                // Remove particles from the old position, and insert it to the
+                // new one. The indexes are valid during the movement, because
+                // we insert a new particle for each particle removed.
+                let particle = self.particles.remove(i);
+                self.particles.insert(new_index, particle);
+
+                // Update molids
+                let _ = self.molids.remove(i);
+                self.molids.insert(new_index, new_molid);
+
+                new_index += 1;
+            }
+        }
+
+        let mut old_mol = self.molecules[old_molid].clone();
+        let size = old_mol.size() as isize;
+
+        // translate all indexes in the molecules between new_mol and old_mol
+        for molecule in self.molecules.iter_mut().skip(new_molid + 1).take(old_molid - new_molid - 1) {
+            molecule.translate_by(size);
+        }
+
+        // Update molid for all particles after the old molecule
+        for molid in self.molids.iter_mut().skip(old_mol.end()) {
+            *molid -= 1;
+        }
+
+        let delta = old_mol.start() - new_mol.end();
+        old_mol.translate_by(- (delta as isize));
+
+        new_mol.merge_with(old_mol);
+        self.molecules[new_molid] = new_mol;
+        let _ = self.molecules.remove(old_molid);
+
+        // Check that self.molids is sorted and only contains successive values
+        #[allow(block_in_if_condition_stmt)]
+        {
+            debug_assert!(self.molids.iter().fold((true, 0), |(is_valid, previous), &i| {
+                if i == previous || i == previous + 1 {
+                    (is_valid, i)
+                } else {
+                    (false, i)
+                }
+            }).0, "Unsorted molecule ids {:?}", self.molids);
+        }
+
+        return delta;
+    }
+}
+
+impl<'a> IntoIterator for &'a Configuration {
```

Comment:
It's subjective, but I don't really like that the system (or configuration) can be iterated. Add `.particles()` in the end is not so long and much more readable.